third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp

```diff
@@ -221,6 +221,78 @@ class TritonAMDGPUReorderInstructionsPass
           dfgop->moveBefore(block, block->begin());
       }
     }
+
+    /**
+     * Sched-load optimization for matmul kernels with large tile sizes
+     * The basic idea of sched-load optimization is to sink the 2nd tt.load
+     * after local_load so that global_load instructions can be interleaved with
+     * mfma's. This can help hide the issue latency of global_load instructions
+     * and improve performance on MI300X.
+     *
+     * It's assumed that the IR before this optimization has the following
+     * structure:
+     * ```mlir
+     * scf.for ..
+     * {
+     *   tileA = tt.load a_ptr
+     *   tileB = tt.load b_ptr
+     *   opA = local_load bufferA
+     *   opB = local_load bufferB
+     *   res = tt.dot opA, opB
+     *   local_store tileA, bufferA
+     *   local_store tileB, bufferB
+     * }
+     * ```
+     * After this optimization, the IR is transformed to
+     * ```mlir
+     * scf.for ..
+     * {
+     *   tileA = tt.load a_ptr
+     *   opA = local_load bufferA
+     *   opB = local_load bufferB
+     *   tileB = tt.load b_ptr  <-- 2nd tt.load is sinked here
+     *   res = tt.dot opA, opB
+     *   local_store tileA, bufferA
+     *   local_store tileB, bufferB
+     * }
+     * ```
+     * For now, we don't have a perfect hueristic about when should this
+     * optimization be applied. Therefore, we implement a simple hueristic that
+     * this is applied when the tile size of A and B are large enough, i.e.
+     * nonKDim >= 128  and kDim >= 64. And also this is only applied for typical
+     * matmul kernels, i.e. only two tt.load's and one dotOp inside the loop. We
+     * are experimenting how to better control instruction scheduling and enable
+     * such optimizations.
+     */
+    m.walk([&](scf::ForOp forOp) -> void {
+      SetVector<Operation *> loadOps;
+      triton::DotOp dotOp;
+      int nDotOps = 0;
+      for (Operation &op : forOp) {
+        if (auto loadOp = dyn_cast<triton::LoadOp>(&op))
+          loadOps.insert(loadOp);
+        if (isa<triton::DotOp>(&op)) {
```

Comment:
Nit: I know I suggested this before but now you can actually do `if (auto curOp = dyn_cast<DotOp(..)) { .. dotOp = curOp; }`.